### PR TITLE
GRPO

### PR DESCRIPTION
**Pull request by winglian**
wip towards adding support for GRPO

### CODE REVIEW

File: src/axolotl/core/trainers/grpo/__init__.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        return grpo_args_kwargs
+
+    @classmethod
+    def set_trainer_kwargs(cls, cfg):
+        trainer_kwargs = {}
+        if cfg.grpo_reward_funcs:
+            reward_funcs = []
+            for reward_func_module in cfg.grpo_reward_funcs:
+                # use importlib to dynamically load the reward function from the module
+                reward_func_module_name = reward_func_module.split(".")[-1]
```
Comment (salmanmohammadi) [inline src/axolotl/core/trainers/grpo/__init__.py]: worth inspecting/asserting the signature of the reward fns?

Comment (winglian) [reply src/axolotl/core/trainers/grpo/__init__.py]: how exactly do you have this in mind?

Comment (winglian) [reply src/axolotl/core/trainers/grpo/__init__.py]: also `RewardFunc` also accepts strings which could be a pretrained model id, so we should probably wrap with a try / except ModuleNotFoundError.

Comment (salmanmohammadi) [reply src/axolotl/core/trainers/grpo/__init__.py]: done!

File: src/axolotl/utils/config/models/input/v0_4_1/trl.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
+"""
+GRPO specific configuration args
+"""
+from typing import List, Optional
+
+from pydantic import BaseModel, Field
+
+
+class TrlConfig(BaseModel):
```
Comment (salmanmohammadi) [inline src/axolotl/utils/config/models/input/v0_4_1/trl.py]: This looks great : )

File: src/axolotl/core/trainers/grpo/trainer.py [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+                        apply_chat_template(x, reward_processing_class)["text"]
+                        for x in messages
+                    ]
+                else:
+                    texts = [p + c for p, c in zip(prompts, completions)]
+                reward_inputs = reward_processing_class(
+                    texts,
+                    return_tensors="pt",
+                    padding=True,
+                    padding_side="right",
```
Comment (abodacs) [inline src/axolotl/core/trainers/grpo/trainer.py]: why there padding_side mismatched between `reward_inputs` and `prompt_inputs`?

File: src/axolotl/utils/trainer.py [hunk 4]
```diff
@@ -576,7 +576,7 @@ def prepare_opinionated_env(cfg):
 def setup_trainer(
     cfg, train_dataset, eval_dataset, model, tokenizer, processor, total_num_steps
 ):
-    if cfg.rl in ("dpo", "ipo", "orpo", "kto", "simpo"):
+    if cfg.rl in ("dpo", "grpo", "ipo", "orpo", "kto", "simpo"):
```
Comment (salmanmohammadi) [inline src/axolotl/utils/trainer.py]: I think we can just check for `if cfg.rl` here right?

File: src/axolotl/core/trainers/grpo/trainer.py (hunk 3 above)
Comment (winglian) [reply src/axolotl/core/trainers/grpo/trainer.py]: this was a copy/paste from upstream trl trainer as I needed this PR to land to simplify our end. https://github.com/huggingface/trl/pull/2817/files I've removed this method from this class now that the referenced PR is merged.

**Comment by winglian** [comment]
I also need https://github.com/axolotl-ai-cloud/axolotl/pull/2193 merged to handle the named config from the dataset we're using for the grpo test.
